Pypp revamp

nilsdeppe opened the pull request:
## Proposed changes

I need the ability to pass `boost::optional`s through pypp, which resulted in a fairly significant revamp of the way arguments are passed to the python code. The major additional improvement is that *any* type can be converted before being passed to the python code. Additionally, these conversions can be defined differently for each translation unit, making the design very flexible and easy to have work across many different tests. This is achieved by replacing the `SliceContainerImpl` struct with a new conversion struct that can be specialized to define how to convert types to and from python types. For example, I use this to pass two different equations of state (polytropic and ideal gas) to python code as a bool `use_polytropic_fluid`. Thus, complex types can be "passed" to the python code without having to write and use python bindings.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Combine Pypp test cases

nilsdeppe committed:
Generalize pypp type conversion for DataVectors

nilsdeppe committed:
Add boost::optional support to pypp

nilsdeppe committed:
Allow locally specifying pypp conversion to python classes

Adds support for specifying how should be manipulated (e.g. get a single grid
point value out) before sending to python. This allows passing complicated
classes, such as an equation of state, into the python code without having to
write bindings or some other general complicated conversion class because the
equation of state can just be converted to a std::array<double> that holds the
values for the equation of state parameters.

nilsdeppe committed:
Add std::optional support to pypp

nilsdeppe committed:
Fix clang-tidy complaints in Pypp

nilsdeppe committed:
Improve error message in Pypp

wthrowe commented:
`ContainerPackAndUnpack` isn't an ODR violation, but isn't `CallImpl<stuff>::call(stuff)` one?  [basic.def.odr] paragraph 6 (mostly 6.2) says "Given [a member function of a class template] defined in more than one translation unit...in each definition...corresponding names...shall refer to the same entity...".  `call` refers to `(anonymous namespace)::ContainerPackAndUnpack`, which I'm pretty sure is a different entity in each translation unit in which it appears because of the anonymous namespace.

nilsdeppe commented:
Hmm yea I interpret that the same way you do. I think moving the `CallImpl` into the anonymous namespace and marking  the free `call` and `call_work` functions static would avoid the issue, right?

wthrowe commented:
That fixes those functions, but then functions in `tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp` and `tests/Unit/Framework/CheckWithRandomValues.hpp` are UB for the same reason.  (`pypp::call` now has internal linkage and so is a different entity in each translation unit.)

nilsdeppe commented:
Hmm, yea anything that calls `pypp` or a wrapping function defined in a header file needs to have internal linkage. I'll try to figure out how big of a change that is.

Do you have any thoughts on other ways of allowing custom conversions of types in only one translation unit? I guess we could do something like a tuple-ish type (maybe TaggedTuple?) that would hold on object of the type that does the translation. My biggest worry about the propagation of internal linkage is that as new code gets added it will inevitably get missed (I think me missing that gives a pretty clear example ;) ).

wthrowe commented:
Yeah, I don't think going the linkage route is a good plan.

I think templating on (or passing an instance of) some type that does the conversion is the right way to go here.

nilsdeppe commented:
Okay, that shouldn't be too difficult to do. I think basically keep the current function names and type aliases, etc., but pass in a `tmpl::list` called `ConvertTypesForPython` and then do a quick check on each argument if that should be used or the generic implementation should be used. Sound reasonable?

wthrowe commented:
This is getting into some complicated rules, so I'm going to just write notes of things that look possibly relevant from looking through the standard and hope they turn into a picture of what is going on.  (I'm skipping a lot of details that I don't think will matter here.)

Linkage:
* Names can have internal or external linkage (or no linkage, but that doesn't look relevant)
* Names with external linkage represent the same entity in all translation units, while those with internal linkage represent different entities in all translation units.
* The names of variables, functions, and function templates have internal linkage if they are static (but not member variables).
* Anything in an unnamed namespace has internal linkage.

The rules for when two names denote the same entity is basically what you would expect.

ODR:
* The description of "odr-use" is a 22 line long paragraph.  I think for our purposes it boils down to: A function is odr-used if it is selected by overload resolution.
* There can be definitions for things with external linkage in multiple translation units if they are classes, inline things, templates, or partial template specializations (but not complete specializations), but they have to follow certain rules:
  * The definitions must be textually the same.
  * Names in each definition must refer to either some part of the thing being defined or the same entity after overload resolution and resolving template specializations.  (For dependent types this check applies only on instantiation.)

For our purposes, I think this all boils down to:
* If an entity with external linkage is defined in more than one translation unit then it cannot refer to anything with internal linkage except when that thing is dependent.
* The previous rule also applies to any instantiated template specializations, in which case no types are considered dependent.

wthrowe commented:
(New comment so GitHub realizes something has changed.)

For this PR, I think this means roughly that, if `f` has external linkage, then for any types `T1...Tn`, `f(T1,...,Tn)` must do the same thing in every translation unit in which it is called.  (Similarly for template parameters.)  I think the easiest way to do that is to encode any custom translations in a (possibly deduced) template parameter.  Any customization a translation unit wants can be defined with internal linkage so there is no way any other translation unit could make a conflicting call.

nilsdeppe committed:
Combine test functions is Test_PyppRandomValues

nilsdeppe committed:
Simplify Test_PyppRandomValues

Use templates and combine checks into single function calls.

nilsdeppe committed:
Simplify CheckWithRandomValues.hpp

nilsdeppe committed:
Simplify CheckWithRandomValues.hpp

nilsdeppe committed:
Simplify CheckWithRandomValues.hpp

No longer require explicitly listing the tags of a tagged tuple when calling
`CheckWithRandomValues`. These can be deduced inside the function.

Add some requires to make the code easier to read.

The extra requires will also make changing the code to support forwarding
arbitrary arguments through easier in the future.

nilsdeppe commented:
Okay, updated this finally...
- rebased on develop
- no longer use anonymous namespace in header. Instead, a `tmpl::list` can be passed to `pypp::call` that defines how to pre-process types. 
- I didn't add the conversion ability to `check_with_random_values` because first those functions need to be rewritten to support passing in non-DataVector/double (or tensors thereof) types. While figuring that out, I did some cleanup and simplifications of the `CheckWithRandomValues` code (e.g. use `if constexpr`). I think a couple more functions could be combined in that file, but that would require touching quite a few of the test files and I don't want to make this PR touch too much code (it already does a lot).

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -21,262 +22,311 @@
-    const size_t npts = get_first_argument(t...).begin()->size();
-    auto return_container = ScalarSpinWeighted{npts};
+    typename ContainerPackAndUnpack<R>::packed_type return_container =
+        ContainerPackAndUnpack<R>::construct_zero_result(npts);
```
Looking at various definitions and uses, it looks like `typename ContainerPackAndUnpack<R>::packed_type` has to always be `R`.  Is there a need for the alias, or at least is there a need for it to be used everywhere?

If there is a reason to use it, it should also be used for the (by reference) return type of the various `pack` functions.

nilsdeppe replied:
It's useful since for custom conversion classes there's no other way (that I've thought of) to retrieve the packed type. I'll make sure it's used the way you suggested.

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -248,6 +248,49 @@ struct ContainerPackAndUnpack<Scalar<SpinWeighted<ValueType, Spin>>,
+    if (UNLIKELY(not static_cast<bool>(t))) {
+      throw std::runtime_error(
+          "Returned type is None in one element of the boost::optional "
+          "container. We can't support this.");
```
I don't understand this message.  An optional only contains one thing, not multiple elements.  And why can't this just return an empty optional?

nilsdeppe replied:
The issue is if one element of a `DataVector` is `None`. I think I can deal with this by only having this check enabled if the `unerlying_packed_type` is a `(Complex)DataVector`. Does that sound okay?

wthrowe replied:
Yes, that makes sense.

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -18,58 +18,40 @@
+                             PackedTypeMatches<tmpl::pin<T>, tmpl::_1>>> {
+  static_assert(
+      tmpl::size<tmpl::find<ConversionClassList,
+                            PackedTypeMatches<tmpl::pin<T>, tmpl::_1>>>::value,
```
I think this is equivalent to the `Requires` check above.  Did you mean a `count_if` or something?

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -291,6 +291,49 @@ struct ContainerPackAndUnpack<boost::optional<T>, std::nullptr_t> {
+    if (UNLIKELY(not static_cast<bool>(t))) {
+      throw std::runtime_error(
+          "Returned type is None in one element of the std::optional "
+          "container. We can't support this.");
```
Same question as for the boost version.

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -426,9 +480,53 @@ struct CallImpl<
+/// - a type alias `packed_type` the type used when converting the result from
+///   the python call back into the C++ code.
+/// - a type alias `underlying_packed_type` which corresponds to the "packed
+///   type" (e.g. DataVector)
```
I don't understand this bullet.

nilsdeppe replied:
I've tried to clarify this. Hopefully I've clarified enough that you can provide a suggested improvement :)

wthrowe replied:
That's clear now, thanks.

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -426,9 +480,53 @@ struct CallImpl<
 ///
 /// Custom classes can be converted between Python and C++ by overloading the
-/// `pypp::ToPythonObject<T>` and `pypp::FromPythonObject<T>` structs for your
+/// `pypp::ToPyObject<T>` and `pypp::FromPyObject<T>` structs for your
```
Is the interface for these documented anywhere?  People can't overload them if they are presented as opaque types.

Maybe just drop support for this and always use the general version below?

nilsdeppe replied:
It is not. I think dropping support for this until it's actually used is fine. At that point we might even want to think about restructuring pypp into a subdirectory and turning it into a more full-fledged library.

wthrowe reviewed (changes requested):
I believe the rewrite has fixed the ODR issues.

nilsdeppe commented:
I've pushed fixups. Some additional changes to try and clarify things:
- renamed `(un)packed_type` to `(un)packed_container` since this type alias really refers to the entire contaire
- renamed `underlying_packed_type` to `packed_type`

Getting the fixups sorted out with the renames took some gymnastics, there was some minor damage done to the `Allow locally specifying pypp conversion to python classes ` that I fixed in the fixup (the final result compiles)

nilsdeppe commented:
@wthrowe friendly ping/reminder that I pushed fixups to this :)

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -21,262 +22,311 @@
-struct convert_to_container_of_doubles<Container<ComplexDataVector, Ts...>> {
-  using type = Container<std::complex<double>, Ts...>;
-};
+  static inline unpacked_type unpack(const DataVector& t,
```
Should `DataVector` -> `packed_type` for consistency with the return value?  (Also for the other specializations.)

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -21,262 +22,311 @@
-struct convert_to_container_of_doubles<
-    Container<SpinWeighted<ComplexDataVector, Spin>, Ts...>> {
-  using type = Container<std::complex<double>, Ts...>;
+  static inline void pack(const gsl::not_null<DataVector*> packed_t,
```
As above, should `DataVector` -> `packed_type`?

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -21,262 +22,311 @@
-         boost::combine(container_double, container_dv)) {
-      boost::get<0>(double_and_datavector_components) =
-          boost::get<1>(double_and_datavector_components)[slice_idx];
+  static inline unpacked_type unpack(const std::array<T, Size>& container_dv,
```
This function does not appear to be DataVector/double specific, so choose different variable names.  I think you were using `t` and `packed_t` above.

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -21,262 +22,311 @@
+    }
+  }
+
+  static inline std::array<T, Size> construct_zero_result(
```
Picking an arbitrary one: Could `make_with_value` be used instead of `construct_zero_result` and `get_size`?

nilsdeppe replied:
My worry would be that some classes like Equations of State don't have a `make_with_value`, but they also wouldn't really be returned from python functions. But even that's not fully supported, so yea, I think switching to `make_with_value` works.

nilsdeppe replied:
`get_size` is used for looping over the grid points, though. I'm not sure how to get around that. `number_of_points` is private (in a detail namespace) and it looks like not required to be implemented for all types(?)

wthrowe reviewed (changes requested):
Sorry about the two new "outdated" comments.  I think if I moved them to the new version of the code GitHub would display them at the bottom, which would be confusing.  I've verified that they are still relevant.

nilsdeppe commented:
@wthrowe I squashed the previous fixups, added new ones, and rebased on develop. I don't know if we can get rid of `get_size` with `make_with_value` since we need some way of getting the number of grid points.

moxcodes commented on tests/Unit/Framework/Tests/Test_PyppAnalyticSolution.cpp:
```diff
@@ -0,0 +1,79 @@
+  const double a = 4.3;
+  const std::array<double, 3> b{{-1.3, 5.6, -0.2}};
+  AnalyticSolutionTest solution{a, b};
+  const DataVector used_for_size(5);
```
(commenting at a somewhat random line that reminds me of this) Would it be difficult to make it possible to call these check functions with the size rather than a `used_for_size`? If possible to get all the type information without it, I think that design would be nicer.

nilsdeppe replied:
Good question! I think it would be possible to basically scan the function arguments passed in and try to identify something that "looks right". I think the random value code needs a bigger overhaul than what I'm giving it in this PR. There are a few ways we could combine more of the functions by using `if constexpr`, and I think it would also be really great to support the `ConversionClassList` business in `check_with_random_values`. I put in a brief effort on that front and found that basically a somewhat larger re-imagining of the implementation would be needed. Probably at the same time as that it could be revisited if there's a way to get rid of the `used_for_size`. I sort of stopped where I did in this PR because it did what I needed, and the next part required a fairly big change. I did keep the easy cleanups to `check_with_random_values` since I had them done. What do you think?

moxcodes replied:
Ah, yeah, okay -- I agree, changing everything to permit overloads or argument inference for just passing the `size_t` would need to change significantly more than the pypp changes in this PR.
Let's leave it for the future, then - - it's definitely true that putting in those alternatives will be easier with the simplified pypp code.

moxcodes commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -175,17 +168,17 @@ struct ContainerPackAndUnpack<std::array<T, Size>, std::nullptr_t> {
 
   static inline unpacked_container unpack(
-      const std::array<T, Size>& container_dv,
+      const packed_container& packed_t,
```
`packed_t` sounds like a type, any chance it could be called something else?

nilsdeppe replied:
good point. Reasonable alternatives I can think of:
1. `packed` and `unpacked`
2. `packed_v` and `unpacked_v`
I would prefer 1, I think. Of course, ideal would be `packed_container` but that I think either won't compile and if it does, would be way too confusing. I'm open to another suggestion too!

moxcodes replied:
Yeah, I like `packed` and `unpacked`

nilsdeppe replied:
Okay, I'll switch to those. Thanks!

moxcodes commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -223,6 +223,53 @@ struct ContainerPackAndUnpack<Scalar<SpinWeighted<ValueType, Spin>>,
+      throw std::runtime_error(
+          "Returned type is None in one element of the boost::optional's "
+          "packed type (DataVector or ComplexDataVector). We can't support "
+          "this because we can just make one element of the packed type be an "
```
`we can just make` -> `we can't just make` (I think...)

moxcodes commented on tests/Unit/Framework/Tests/Test_Pypp.cpp:
```diff
@@ -411,6 +412,35 @@ void test_function_of_time() {
+        pypp::call<T>("PyppPyTests", "add_scalars", a, Optional<T>{});
+    CHECK_ITERABLE_APPROX(get(result_double_none), get(*a));
+
+    const auto result_none_double =
```
the `none_none` case appears to be in the pypp py part, but is not tested in the test.

nilsdeppe replied:
Good catch! This would require adding of `make_with_value`s for `*::optional<T>` and also handling `optional<T>` at the `ToPyObject` level and `FromPyObject` level. I'm going to assume we aren't going to be returning a `None` from the python code, though support I think could be added in the future. Part of the issue is that we choose the `CallImpl` implicitly based on the return type, so returning `optional<T>` is not so easy right now.  Probably the best way to handle this would be to have a `call_over_grid` function in addition to the `call` function already in `pypp`. We then explicitly know whether we need to loop or not. I'd prefer to defer that level of change, though since that'll propagate to more than 100 targets/files.

Edit: I'm removing the unused case in the python file

moxcodes commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -26,58 +26,40 @@
-    throw std::runtime_error{"Function returned null"};
-  }
+template <typename T, typename ConversionClass>
+struct PackedTypeMatches : std::false_type {};
```
is it viable to just create a type alias for `std::is_same_v<>` (i.e. can you assume that all of the conversion classes have the desired type alias)?

nilsdeppe replied:
Yes, that's part of the required interface so it should work. I'll give it a shot and report back if I figured out a reason it doesn't :)

moxcodes commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -26,58 +26,40 @@
-template <>
-struct ContainerPackAndUnpack<DataVector, std::nullptr_t> {
+template <typename T, typename ConversionClassList>
+struct ContainerPackAndUnpack<
```
What's the design reasoning for making this take an arbitrary conversion class list, rather than just asking new pack routines specialize the `ContainerPackAndUnpack` struct?

nilsdeppe replied:
ODR violations... There is some discussion above on this PR and a very nice comment by Will that explains that the issue with the specializations is that then every test file needs to define a unique specialization. However, I could reasonable see someone wanting to convert the same type (e.g. an EOS) but with different parameters in different files. E.g. I want a polytropic EOS with gamma=4/3 in file A and gamma=5/3 in file B. With different specializations in different files this leads to ODR violations. Even trying to fix this by marking some of the wrapping classes and functions as having internal linkage, this procedures remains brittle and errors will be impossible to debug for anyone except for the people who've had this conversation (currently 3 of us). Passing in the type list makes it so that you can specify any conversion in any local file and there will not be any weird C++ funny business. Seem alright?

moxcodes replied:
Aha, understood. I suppose ODR violations are still _technically_ possible here, though, right? If two different tests define a new class `struct EosPackAndUnpack` with different contents, and have identical conversion class lists (`tmpl::list<EosPackAndUnpack>`), then the same problem comes up, does it not?
(I'm not advocating for any further indirection, I'm just trying to come to a working grasp of how avoiding the ODR violation works)

nilsdeppe replied:
Yea, this is a bit tricky. The key thing is what @wthrowe said:

> For this PR, I think this means roughly that, if f has external linkage, then for any types T1...Tn, f(T1,...,Tn) must do the same thing in every translation unit in which it is called. (Similarly for template parameters.) I think the easiest way to do that is to encode any custom translations in a (possibly deduced) template parameter. Any customization a translation unit wants can be defined with internal linkage so there is no way any other translation unit could make a conflicting call.

This means that by making `pypp::call` explicitly depend on the locally defined (with internal linkage) struct we get a new definition of the function in the translation unit. Adding a specialization of the `ContainerPackAndUnpack` in a translation unit is not an explicit dependency of `pypp::call` and therefore you have one definition for all translation units. This is where the ODR violation comes in. It's straightforward to create a little example:
```
 $ cat test2.cpp    
#include <vector>
#include <memory>
#include <iostream>


void foo_a();
void foo_b();

int main() {
  foo_a();
  foo_b();
}

 $ cat test.hpp  
#pragma once

#include <iostream>

template <typename T>
struct Printer {
  void print() {
    std::cout << "Hello\n";
  }
};

template <typename T>
void printer() {
  T::print();
}

 $ cat fooa.cpp
#include <iostream>

#include "test.hpp"


namespace {
struct MyPrinter {
  static void print() {
    std::cout << "Hi from a\n";
  }
};
}  // namespace

template <>
struct Printer<double> {
  static void print() {
    std::cout << "Hello from a\n";
  }
};

void foo_a() {
  printer<MyPrinter>();
  printer<Printer<double>>();
}

 $ cat foob.cpp
#include <iostream>

#include "test.hpp"


namespace {
struct MyPrinter {
  static void print() {
    std::cout << "Hi from b\n";
  }
};
}  // namespace

template <>
struct Printer<double> {
  static void print() {
    std::cout << "Hello from b\n";
  }
};

void foo_b() {
  printer<MyPrinter>();
  printer<Printer<double>>();
}
```
The output is:
```
Hi from a
Hello from a
Hi from b
Hello from a
```

moxcodes commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -21,262 +22,309 @@
 /// Contains all functions for calling python from C++
 namespace pypp {
 namespace detail {
+template <typename R>
```
I'd prefer a more explicit template parameter (also there are a few others like that in this file, but for some reason `R` seems more confusing to me than `T`)

nilsdeppe replied:
Ah, `R` is short for `ReturnType`. How about I add a commit that makes the change in the entire file at the end?

moxcodes replied:
Sounds great

nilsdeppe committed:
Rename R to ReturnType in pypp

nilsdeppe commented:
Rebased on develop and pushed new fixup commits. I also added a commit that renames `R` to `ReturnType`

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -81,21 +81,22 @@ struct ContainerPackAndUnpack<DataVector, std::nullptr_t> {
+      const packed_container& packed, const size_t grid_point_index) noexcept {
+    ASSERT(grid_point_index < packed.size(),
+           "Trying to slice DataVector of size " << packed.size()
+                                                 << "with grid_point_index "
```
Missing space.

wthrowe commented on tests/Unit/Framework/Pypp.hpp:
```diff
@@ -106,21 +107,22 @@ struct ContainerPackAndUnpack<ComplexDataVector, std::nullptr_t> {
            "Trying to slice ComplexDataVector of size "
-               << t.size() << "with grid_point_index " << grid_point_index);
-    return t[grid_point_index];
+               << packed.size() << "with grid_point_index "
```
Missing space.

wthrowe reviewed (changes requested):
Just a couple of trivial changes.  Squash directly.

nilsdeppe commented:
@moxcodes let me know when you've looked at the fixups and I can squash :)

moxcodes commented:
looks good, thanks. Go ahead and squash

nilsdeppe commented:
Okay, squashed, including adding the missing spaces. I went through and built and ran tests for each commit on clang-11 and gcc-7 to be sure nothing is obviously broken. Thanks for the reviews!
